File: src/main/java/no/mesan/fag/patterns/timesheet/command/AsyncTaskExecutor.java
package no.mesan.fag.patterns.timesheet.command;

import java.util.Random;
import java.util.logging.Logger;

/** Kjører kommandoer på oppfordring. */
public final class AsyncTaskExecutor {

    private static final Logger LOG = Logger.getLogger(AsyncTaskExecutor.class.getName());
    private static final Random random = new Random();

    /**
     * Kjør et antall oppgaver.
     * @param tasks
     */
	public void executeTasks(final AsyncTask ... tasks) {
        for (AsyncTask task : tasks) {
            final Thread thread = new Thread(new AsyncTaskThreadRunner(task));
    		thread.start();
        }
	}

    /** Her skjer selve eksekveringen. */
	private class AsyncTaskThreadRunner implements Runnable {

		private final AsyncTask task;

		AsyncTaskThreadRunner(final AsyncTask task) {
			this.task = task;
		}

		@Override
		public void run() {
			LOG.info("Running background task " + task.whoAmI() + "...");
			try {
				Thread.sleep(random.nextInt(2000)+500);
			}
            catch (final InterruptedException e) {
				e.printStackTrace(); // Ikke si det til noen
			}
            task.executeTask();
			LOG.info("Finished background task " + task.whoAmI());
		}
	}
}

File: src/main/scala/no/mesan/fag/patterns/scala/timesheet/command/AsyncTaskExecutor.scala
package no.mesan.fag.patterns.scala.timesheet.command

import java.util.logging.Logger
import scala.util.Random

/** Kjører kommandoer på oppfordring. */
class AsyncTaskExecutor {
  private val Log = Logger.getLogger(classOf[AsyncTaskExecutor].getName)
  private val random= Random

  /**
   * Kjør et antall oppgaver.
   * @param tasks oppgaveliste
   */
  def executeTasks(tasks: AsyncTask*) = for (task <- tasks)
    new Thread(new AsyncTaskThreadRunner(task)).start

  /** Her skjer selve eksekveringen. */
  private class AsyncTaskThreadRunner(task: AsyncTask) extends Runnable {
    override def run {
      Log.info("Running background task " + task.whoAmI + "...")
      Thread.sleep(random.nextInt(2000) + 500)
      task.executeTask
      Log.info("Finished background task " + task.whoAmI)
    }
  }
}
